spatialos-sdk/src/worker/worker_future.rs

```diff
@@ -0,0 +1,91 @@
+use futures::channel::oneshot::*;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    thread,
+    thread::JoinHandle,
+};
+
+pub struct WorkerFuture<T: WorkerSdkFuture> {
+    inner: Option<WorkerFutureState<T>>,
+}
+
+impl<T: WorkerSdkFuture> WorkerFuture<T> {
+    pub fn new(future: T) -> Self {
+        WorkerFuture {
+            inner: Some(WorkerFutureState::NotStarted(future)),
+        }
+    }
+}
+
+enum WorkerFutureState<T: WorkerSdkFuture> {
+    NotStarted(T),
+    InProgress(InProgressHandle<T>),
+}
+
+struct InProgressHandle<T: WorkerSdkFuture> {
+    pub result_rx: Receiver<T::Output>,
+    pub thread: Option<JoinHandle<()>>,
+}
+
+impl<T: WorkerSdkFuture> Drop for WorkerFuture<T> {
+    fn drop(&mut self) {
+        if let Some(future_state) = &mut self.inner {
+            if let WorkerFutureState::InProgress(handle) = future_state {
+                handle.thread.take().unwrap().join().unwrap();
+            }
+        }
+    }
+}
+
+pub trait WorkerSdkFuture: Send + Unpin + 'static {
+    type RawPointer;
+    type Output: Send;
+
+    fn start(&self) -> *mut Self::RawPointer;
+    unsafe fn get(ptr: *mut Self::RawPointer) -> Self::Output;
+    unsafe fn destroy(ptr: *mut Self::RawPointer);
+}
+
+impl<T: WorkerSdkFuture> Future for WorkerFuture<T> {
+    type Output = T::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let inner = &mut self.as_mut().get_mut().inner;
+        // This panics if the future is polled after completion. This is acceptable according to
+        // the Future trait contract - https://doc.rust-lang.org/std/future/trait.Future.html#panics
+        let mut future_state = inner.take().unwrap();
+
+        match future_state {
+            WorkerFutureState::NotStarted(ftr) => {
+                let (tx, rx) = futures::channel::oneshot::channel();
+                let task = cx.waker().clone();
+
+                let thread = thread::spawn(move || unsafe {
+                    let ptr = ftr.start();
+                    let value = T::get(ptr);
```

Comment:
Would it make sense to fuse `start` and `get` into a single function? Is there any advantage to having them be separate at this point?

Reply:
If we want to add options like in https://github.com/jamiebrynes7/spatialos-sdk-rs/issues/137 it makes sense to keep them separate

Reply:
How do you mean? I would think that having a single, safe method would make it easier for users to manually poll the future if they prefer to do so.

Reply:
Well currently the `start()` method creates the underlying C future, and `get` polls that future. 

So the code above mapped to a C API future is something like: 

```
Worker_ConnectionFuture* ptr = Worker_ConnectionFuture_Create();
Worker_Connection* conn = Worker_ConnectionFuture_Get(ptr);
```

These can't be combined into the same method if you want to call `Get` more than once. Right now this is true, but if we open up different polling options then it's not true.

Reply:
Ah, yeah I see what you mean. We'll probably want to give the whole future setup another pass in order to make it more flexible, but the current version seems fine for now 👍